### PR TITLE
dont catch dynamic and force Exception

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description
The catch here catches any thrown type as dynamic, but passes it to a function that expects an Exception - this is unsound and unnecessary. Widen throwToolExitIfLicenseNotAccepted to accept any type..

```
_TypeError: type 'StateError' is not a subtype of type 'Exception

  | at _readGradleProject | (gradle.dart:260 )
-- | -- | --
  | at <asynchronous gap> | (async )
  | at GradleUtils.getAppProject | (gradle.dart:55 )
  | at <asynchronous gap> | (async )
  | at getGradleAppOut | (gradle.dart:143 )
  | at <asynchronous gap> | (async )
  | at AndroidApk.fromAndroidProject | (application_package.dart:168 )
  | at <asynchronous gap> | (async )
  | at ApplicationPackageFactory.getPackageForPlatform | (application_package.dart:47 )
  | at <asynchronous gap> | (async )
  | at FlutterDevice.runHot | (resident_runner.dart:357 )
  | at <asynchronous gap> | (async )
  | at HotRunner.run | (run_hot.dart:258 )
  | at <asynchronous gap> | (async )
  | at RunCommand.runCommand | (run.dart:481 )
  | at <asynchronous gap> | (async )
  | at FlutterCommand.verifyThenRunCommand | (flutter_command.dart:553 )
  | at _asyncThenWrapperHelper.<anonymous closure> | (async_patch.dart:71 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._completeWithValue | (future_impl.dart:522 )
  | at _AsyncAwaitCompleter.complete | (async_patch.dart:30 )
  | at _completeOnAsyncReturn | (async_patch.dart:288 )
  | at RunCommand.usageValues | (run.dart )
  | at _asyncThenWrapperHelper.<anonymous closure> | (async_patch.dart:71 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._completeWithValue | (future_impl.dart:522 )
  | at _AsyncAwaitCompleter.complete | (async_patch.dart:30 )
  | at _completeOnAsyncReturn | (async_patch.dart:288 )
  | at IosProject.isSwift | (project.dart )
  | at _asyncThenWrapperHelper.<anonymous closure> | (async_patch.dart:71 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._completeWithValue | (future_impl.dart:522 )
  | at _AsyncAwaitCompleter.complete | (async_patch.dart:30 )
  | at _completeOnAsyncReturn | (async_patch.dart:288 )
  | at IosProject.buildSettings | (project.dart )
  | at _asyncThenWrapperHelper.<anonymous closure> | (async_patch.dart:71 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._completeWithValue | (future_impl.dart:522 )
  | at _AsyncAwaitCompleter.complete | (async_patch.dart:30 )
  | at _completeOnAsyncReturn | (async_patch.dart:288 )
  | at XcodeProjectInterpreter.getBuildSettings | (xcodeproj.dart )
  | at _asyncThenWrapperHelper.<anonymous closure> | (async_patch.dart:71 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._completeWithValue | (future_impl.dart:522 )
  | at _AsyncAwaitCompleter.complete | (async_patch.dart:30 )
  | at _completeOnAsyncReturn | (async_patch.dart:288 )
  | at _DefaultProcessUtils.run | (process.dart )
  | at _asyncThenWrapperHelper.<anonymous closure> | (async_patch.dart:71 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._completeWithValue | (future_impl.dart:522 )
  | at Future.wait.<anonymous closure> | (future.dart:400 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._complete | (future_impl.dart:512 )
  | at _BufferingStreamSubscription.asFuture.<anonymous closure> | (stream_impl.dart:204 )
  | at _rootRun | (zone.dart:1120 )
  | at _CustomZone.run | (zone.dart:1021 )
  | at _CustomZone.runGuarded | (zone.dart:923 )
  | at _BufferingStreamSubscription._sendDone.sendDone | (stream_impl.dart:389 )
  | at _BufferingStreamSubscription._sendDone | (stream_impl.dart:399 )
  | at _BufferingStreamSubscription._close | (stream_impl.dart:283 )
  | at _SinkTransformerStreamSubscription._close | (stream_transformers.dart:96 )
  | at _EventSinkWrapper.close | (stream_transformers.dart:23 )
  | at _StringAdapterSink.close | (string_conversion.dart:249 )
  | at _Utf8ConversionSink.close | (string_conversion.dart:300 )
  | at _ConverterStreamEventSink.close | (chunked_conversion.dart:80 )
  | at _SinkTransformerStreamSubscription._handleDone | (stream_transformers.dart:141 )
  | at _rootRun | (zone.dart:1120 )
  | at _CustomZone.run | (zone.dart:1021 )
  | at _CustomZone.runGuarded | (zone.dart:923 )
  | at _BufferingStreamSubscription._sendDone.sendDone | (stream_impl.dart:389 )
  | at _BufferingStreamSubscription._sendDone | (stream_impl.dart:399 )
  | at _BufferingStreamSubscription._close | (stream_impl.dart:283 )
  | at _SyncStreamControllerDispatch._sendDone | (stream_controller.dart:772 )
  | at _StreamController._closeUnchecked | (stream_controller.dart:629 )
  | at _StreamController.close | (stream_controller.dart:622 )
  | at _Socket._onData | (socket_patch.dart:1836 )
  | at _rootRunUnary | (zone.dart:1136 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _CustomZone.runUnaryGuarded | (zone.dart:931 )
  | at _BufferingStreamSubscription._sendData | (stream_impl.dart:336 )
  | at _BufferingStreamSubscription._add | (stream_impl.dart:263 )
  | at _SyncStreamControllerDispatch._sendData | (stream_controller.dart:764 )
  | at _StreamController._add | (stream_controller.dart:640 )
  | at _StreamController.add | (stream_controller.dart:586 )
  | at new _RawSocket.<anonymous closure> | (socket_patch.dart:1384 )
  | at _NativeSocket.issueReadEvent.issue | (socket_patch.dart:890 )
  | at _microtaskLoop | (schedule_microtask.dart:41 )
  | at _startMicrotaskLoop | (schedule_microtask.dart:50 )
  | at _runPendingImmediateCallback | (isolate_patch.dart:116 )
  | at _RawReceivePortImpl._handleMessage | (isolate_patch.dart:173 )
```

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I haven't added a test because I think this is essentially a type error that isn't caught due to dynamic.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 1]
```diff
@@ -253,7 +253,7 @@ Future<GradleProject> _readGradleProject(
       environment: gradleEnv,
     );
     project = GradleProject.fromAppProperties(propertiesRunResult.stdout, tasksRunResult.stdout);
-  } catch (exception) {
+  } on Exception catch (exception) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/gradle.dart]: It looks like this should only be catching `ProcessException`. We'd want a crash report sent to let us know that we are failing to parse `propertiesRunResult.stdout`. I suspect that is what leads to the `StateError` here, but the problem is obscured by the `TypeError`.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 2]
```diff
@@ -253,7 +253,7 @@ Future<GradleProject> _readGradleProject(
...
     if (getFlutterPluginVersion(flutterProject.android) == FlutterPluginVersion.managed) {
       status.cancel();
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Cleaning up `status` should go in a `finally {}`.

**Comment by jonahwilliams** [comment]
This got really out of sync, will try again later

**Pull request closed without merging**